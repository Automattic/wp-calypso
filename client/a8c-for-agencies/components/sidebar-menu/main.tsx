import { category, home, plugins, tag, currencyDollar } from '@wordpress/icons';
import { useTranslate } from 'i18n-calypso';
import { useMemo } from 'react';
import Sidebar from '../sidebar';
import {
	A4A_OVERVIEW_LINK,
	A4A_PLUGINS_LINK,
	A4A_SITES_LINK,
	A4A_MARKETPLACE_LINK,
	A4A_PURCHASES_LINK,
	A4A_LICENSES_LINK,
} from './lib/constants';
import { createItem } from './lib/utils';

type Props = {
	path: string;
};

export default function ( { path }: Props ) {
	const translate = useTranslate();
	const menuItems = useMemo( () => {
		return [
			{
				icon: home,
				path: '/',
				link: A4A_OVERVIEW_LINK,
				title: translate( 'Overview' ),
				trackEventProps: {
					menu_item: 'Automattic for Agencies / Overview',
				},
			},
			{
				icon: category,
				path: '/',
				link: A4A_SITES_LINK,
				title: translate( 'Sites' ),
				trackEventProps: {
					menu_item: 'Automattic for Agencies / Dashboard',
				},
				withChevron: true,
			},
			{
				icon: plugins,
				path: '/',
				link: A4A_PLUGINS_LINK,
				title: translate( 'Plugins' ),
				trackEventProps: {
					menu_item: 'Automattic for Agencies / Plugins',
				},
			},
			{
				icon: tag,
				path: '/',
				link: A4A_MARKETPLACE_LINK,
				title: translate( 'Marketplace' ),
				trackEventProps: {
					menu_item: 'Automattic for Agencies / Marketplace',
				},
			},
			{
				icon: currencyDollar,
				path: A4A_PURCHASES_LINK,
				link: A4A_LICENSES_LINK,
				title: translate( 'Purchases' ),
				trackEventProps: {
					menu_item: 'Automattic for Agencies / Purchases',
				},
				withChevron: true,
			},
		].map( ( item ) => createItem( item, path ) );
	}, [ path, translate ] );

	return <Sidebar path="" menuItems={ menuItems } withUserProfileFooter />;
}
