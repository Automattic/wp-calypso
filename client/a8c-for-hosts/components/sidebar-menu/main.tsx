import Sidebar from '../sidebar';
import useMainMenuItems from './hooks/use-main-menu-items';

type Props = {
	path: string;
};

export default function ( { path }: Props ) {
	const menuItems = useMainMenuItems( path );

	return <Sidebar path="" menuItems={ menuItems } withUserProfileFooter />;
}
