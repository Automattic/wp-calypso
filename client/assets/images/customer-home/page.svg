<svg width="32" height="32" viewBox="0 0 32 32" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M18.6673 2.66602H8.00065C6.53398 2.66602 5.34732 3.86602 5.34732 5.33268L5.33398 26.666C5.33398 28.1327 6.52065 29.3327 7.98732 29.3327H24.0006C25.4673 29.3327 26.6673 28.1327 26.6673 26.666V10.666L18.6673 2.66602ZM24.0006 26.666H8.00065V5.33268H17.5607L24.0006 11.7727V26.666Z" fill="#2C3338"/>
<path d="M15.3335 15.332V21.9987" stroke="#2C3338" stroke-width="3" stroke-linecap="square"/>
<path d="M18.6665 18.666H11.9998" stroke="#2C3338" stroke-width="3" stroke-linecap="square"/>
</svg>
