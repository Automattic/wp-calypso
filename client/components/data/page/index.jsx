/**
 * External dependencies.
 */
import { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import each from 'lodash/each';

/**
 * Internal dependencies.
 */
import {
	setTitle,
	setDescription,
	addLink,
	addMeta,
	setUnreadCount
} from 'state/page/actions';

class Page extends Component {
	componentWillMount() {
		this.props.setTitle( this.props.title );
		this.props.setDescription( this.props.description );
		this.props.setUnreadCount( this.props.unreadCount );

		if ( this.props.link && this.props.link.length ) {
			each( this.props.link, ( link ) => {
				this.props.addLink( link );
			} );
		}

		if ( this.props.meta && this.props.meta.length ) {
			each( this.props.meta, ( meta ) => {
				this.props.addMeta( meta );
			} );
		}
	}

	componentWillReceiveProps( nextProps ) {
		if ( this.props.title !== nextProps.title ) {
			this.props.setTitle( nextProps.title );
		}

		if ( this.props.description !== nextProps.description ) {
			this.props.setDescription( nextProps.description );
		}

		if ( this.props.unreadCount !== nextProps.unreadCount ) {
			this.props.setUnreadCount( nextProps.unreadCount );
		}

		if ( this.props.link !== nextProps.link && this.props.link.length ) {
			each( nextProps.link, ( link ) => {
				this.props.addLink( link );
			} );
		}

		if ( this.props.meta !== nextProps.meta && this.props.meta.length ) {
			each( nextProps.meta, ( meta ) => {
				this.props.addMeta( meta );
			} );
		}
	}

	render() {
		return null;
	}
}

Page.propTypes = {
	title: PropTypes.string,
	description: PropTypes.string,
	unreadCount: PropTypes.number,
	link: PropTypes.array,
	meta: PropTypes.array,

	setTitle: PropTypes.func.isRequired,
	setDescription: PropTypes.func.isRequired,
	addLink: PropTypes.func.isRequired,
	addMeta: PropTypes.func.isRequired,
	setUnreadCount: PropTypes.func.isRequired
};

Page.defaultProps = {
	title: '',
	unreadCount: 0,
	link: [],
	meta: [],
	description: ''
};

export default connect( null, {
	setTitle,
	setDescription,
	addLink,
	addMeta,
	setUnreadCount
} )( Page );
