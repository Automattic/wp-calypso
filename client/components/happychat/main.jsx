/** @format */

/**
 * External dependencies
 */
import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

/**
 * Internal dependencies
 */
import { HappychatConnection } from './connection';
import { Title } from './title';
import { Composer } from './composer';
import { Notices } from './notices';
import { Timeline } from './timeline';

/*
 * Main chat UI component
 */
export class Happychat extends React.Component {
	componentDidMount() {
		this.props.onFocused();
	}

	componentWillUnmount() {
		this.props.onBlurred();
	}

	render() {
		const {
			chatStatus,
			connectionStatus,
			currentUserEmail,
			disabled,
			getAuth,
			isChatOpen,
			isConnectionUninitialized,
			isCurrentUser,
			isHappychatEnabled,
			isMinimizing,
			isServerReachable,
			message,
			onCloseChat,
			onInitConnection,
			onMinimizeChat,
			onMinimizedChat,
			onSendMessage,
			onSendNotTyping,
			onSendTyping,
			onSetCurrentMessage,
			timeline,
			translate,
		} = this.props;

		return (
			<div className="happychat">
				<HappychatConnection
					getAuth={ getAuth }
					onInitConnection={ onInitConnection }
					isHappychatEnabled={ isHappychatEnabled }
					isConnectionUninitialized={ isConnectionUninitialized }
				/>
				<div
					className={ classnames( 'happychat__container', {
						'is-open': isChatOpen,
						'is-minimizing': isMinimizing,
					} ) }
				>
					<Title
						onCloseChat={ onCloseChat }
						onMinimizeChat={ onMinimizeChat }
						onMinimizedChat={ onMinimizedChat }
						translate={ translate }
					/>
					<Timeline
						currentUserEmail={ currentUserEmail }
						isCurrentUser={ isCurrentUser }
						timeline={ timeline }
						translate={ translate }
					/>
					<Notices
						chatStatus={ chatStatus }
						connectionStatus={ connectionStatus }
						isServerReachable={ isServerReachable }
						translate={ translate }
					/>
					<Composer
						disabled={ disabled }
						message={ message }
						onSendMessage={ onSendMessage }
						onSendNotTyping={ onSendNotTyping }
						onSendTyping={ onSendTyping }
						onSetCurrentMessage={ onSetCurrentMessage }
						translate={ translate }
					/>
				</div>
			</div>
		);
	}
}

Happychat.propTypes = {
	chatStatus: PropTypes.string,
	connectionStatus: PropTypes.string,
	currentUserEmail: PropTypes.string,
	disabled: PropTypes.bool,
	getAuth: PropTypes.func,
	isChatOpen: PropTypes.bool,
	isConnectionUninitialized: PropTypes.bool,
	isCurrentUser: PropTypes.func,
	isHappychatEnabled: PropTypes.bool,
	isMinimizing: PropTypes.bool,
	isServerReachable: PropTypes.bool,
	message: PropTypes.string,
	onBlurred: PropTypes.func,
	onCloseChat: PropTypes.func,
	onFocused: PropTypes.func,
	onInitConnection: PropTypes.func,
	onMinimizeChat: PropTypes.func,
	onMinimizedChat: PropTypes.func,
	onSendMessage: PropTypes.func,
	onSendNotTyping: PropTypes.func,
	onSendTyping: PropTypes.func,
	onSetCurrentMessage: PropTypes.func,
	timeline: PropTypes.array,
	translate: PropTypes.func,
};
