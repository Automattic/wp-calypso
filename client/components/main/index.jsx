/** @ssr-ready **/

/**
 * External dependencies
 */
import React from 'react';
import classNames from 'classnames';

export default function Main( {
	className,
	children,
	wideLayout = false
} ) {
	const classes = classNames( className, 'main', {
		'is-wide-layout': wideLayout
	} );

	return (
		<main className={ classes } role="main">
			{ children }
		</main>
	);
}
