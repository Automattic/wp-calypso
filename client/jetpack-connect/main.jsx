/** @format */
/**
 * External dependencies
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Gridicon from 'gridicons';
import { flowRight, includes } from 'lodash';
import { localize } from 'i18n-calypso';

/**
 * Internal dependencies
 */
import Button from 'components/button';
import Card from 'components/card';
import LoggedOutFormLinks from 'components/logged-out-form/links';
import LoggedOutFormLinkItem from 'components/logged-out-form/link-item';
import JetpackConnectNotices from './jetpack-connect-notices';
import SiteUrlInput from './site-url-input';
import { getConnectingSite, getJetpackSiteByUrl } from 'state/jetpack-connect/selectors';
import { isRequestingSites } from 'state/sites/selectors';
import JetpackInstallStep from './install-step';
import versionCompare from 'lib/version-compare';
import LocaleSuggestions from 'components/locale-suggestions';
import { recordTracksEvent } from 'state/analytics/actions';
import MainWrapper from './main-wrapper';
import FormattedHeader from 'components/formatted-header';
import HelpButton from './help-button';
import JetpackConnectHappychatButton from './happychat-button';
import untrailingslashit from 'lib/route/untrailingslashit';
import { retrievePlan } from './persistence-utils';
import {
	confirmJetpackInstallStatus,
	dismissUrl,
	goToRemoteAuth,
	goToPluginInstall,
	goToPlans,
	goToPluginActivation,
	checkUrl,
} from 'state/jetpack-connect/actions';

/**
 * Constants
 */
const MINIMUM_JETPACK_VERSION = '3.9.6';

class JetpackConnectMain extends Component {
	static propTypes = {
		locale: PropTypes.string,
		path: PropTypes.string,
		type: PropTypes.oneOf( [ 'install', 'pro', 'premium', 'personal', false ] ),
		url: PropTypes.string,
	};

	state = this.props.url
		? {
				currentUrl: this.cleanUrl( this.props.url ),
				shownUrl: this.props.url,
				waitingForSites: false,
			}
		: {
				currentUrl: '',
				shownUrl: '',
				waitingForSites: false,
			};

	componentWillMount() {
		if ( this.props.url ) {
			this.checkUrl( this.cleanUrl( this.props.url ) );
		}
	}

	componentDidMount() {
		let from = 'direct';
		if ( this.props.type === 'install' ) {
			from = 'jpdotcom';
		}
		if ( this.props.type === 'pro' ) {
			from = 'ad';
		}
		if ( this.props.type === 'premium' ) {
			from = 'ad';
		}
		if ( this.props.type === 'personal' ) {
			from = 'ad';
		}
		this.props.recordTracksEvent( 'calypso_jpc_url_view', {
			jpc_from: from,
		} );
	}

	componentDidUpdate() {
		if (
			this.getStatus() === 'notConnectedJetpack' &&
			this.isCurrentUrlFetched() &&
			! this.props.jetpackConnectSite.isRedirecting
		) {
			return this.props.goToRemoteAuth( this.state.currentUrl );
		}
		if ( this.getStatus() === 'alreadyOwned' && ! this.props.jetpackConnectSite.isRedirecting ) {
			return this.props.goToPlans( this.state.currentUrl );
		}

		if ( this.state.waitingForSites && ! this.props.isRequestingSites ) {
			// eslint-disable-next-line react/no-did-update-set-state
			this.setState( { waitingForSites: false } );
			this.checkUrl( this.state.currentUrl );
		}
	}

	dismissUrl = () => this.props.dismissUrl( this.state.currentUrl );

	isCurrentUrlFetched() {
		return (
			this.props.jetpackConnectSite &&
			this.state.currentUrl === this.props.jetpackConnectSite.url &&
			this.props.jetpackConnectSite.isFetched
		);
	}

	isCurrentUrlFetching() {
		return (
			this.state.currentUrl !== '' &&
			this.props.jetpackConnectSite &&
			this.state.currentUrl === this.props.jetpackConnectSite.url &&
			this.props.jetpackConnectSite.isFetching
		);
	}

	handleUrlChange = event => {
		const url = event.target.value;
		this.setState( {
			currentUrl: this.cleanUrl( url ),
			shownUrl: url,
		} );
	};

	cleanUrl( inputUrl ) {
		let url = inputUrl.trim().toLowerCase();
		if ( url && url.substr( 0, 4 ) !== 'http' ) {
			url = 'http://' + url;
		}
		return untrailingslashit( url );
	}

	checkUrl( url ) {
		return this.props.checkUrl( url, !! this.props.getJetpackSiteByUrl( url ), this.props.type );
	}

	handleUrlSubmit = () => {
		this.props.recordTracksEvent( 'calypso_jpc_url_submit', {
			jetpack_url: this.state.currentUrl,
		} );
		if ( this.props.isRequestingSites ) {
			this.setState( { waitingForSites: true } );
		} else {
			this.checkUrl( this.state.currentUrl );
		}
	};

	installJetpack = () => {
		this.props.recordTracksEvent( 'calypso_jpc_instructions_click', {
			jetpack_funnel: this.state.currentUrl,
			type: 'install_jetpack',
		} );

		this.props.goToPluginInstall( this.state.currentUrl );
	};

	activateJetpack = () => {
		this.props.recordTracksEvent( 'calypso_jpc_instructions_click', {
			jetpack_funnel: this.state.currentUrl,
			type: 'activate_jetpack',
		} );
		this.props.goToPluginActivation( this.state.currentUrl );
	};

	checkProperty( propName ) {
		return (
			this.state.currentUrl &&
			this.props.jetpackConnectSite &&
			this.props.jetpackConnectSite.data &&
			this.isCurrentUrlFetched() &&
			this.props.jetpackConnectSite.data[ propName ]
		);
	}

	isRedirecting() {
		return (
			this.props.jetpackConnectSite &&
			this.props.jetpackConnectSite.isRedirecting &&
			this.isCurrentUrlFetched()
		);
	}

	getStatus() {
		if ( this.state.currentUrl === '' ) {
			return false;
		}

		if ( this.checkProperty( 'userOwnsSite' ) ) {
			return 'alreadyOwned';
		}

		if ( this.props.jetpackConnectSite.installConfirmedByUser === false ) {
			return 'notJetpack';
		}

		if ( this.props.jetpackConnectSite.installConfirmedByUser === true ) {
			return 'notActiveJetpack';
		}

		if (
			this.state.currentUrl.toLowerCase() === 'http://wordpress.com' ||
			this.state.currentUrl.toLowerCase() === 'https://wordpress.com'
		) {
			return 'wordpress.com';
		}
		if ( this.checkProperty( 'isWordPressDotCom' ) ) {
			return 'isDotCom';
		}
		if ( ! this.checkProperty( 'exists' ) ) {
			return 'notExists';
		}
		if ( ! this.checkProperty( 'isWordPress' ) ) {
			return 'notWordPress';
		}
		if ( ! this.checkProperty( 'hasJetpack' ) ) {
			return 'notJetpack';
		}
		const jetpackVersion = this.checkProperty( 'jetpackVersion' );
		if ( jetpackVersion && versionCompare( jetpackVersion, MINIMUM_JETPACK_VERSION, '<' ) ) {
			return 'outdatedJetpack';
		}
		if ( ! this.checkProperty( 'isJetpackActive' ) ) {
			return 'notActiveJetpack';
		}
		if (
			! this.checkProperty( 'isJetpackConnected' ) ||
			( this.checkProperty( 'isJetpackConnected' ) && ! this.checkProperty( 'userOwnsSite' ) )
		) {
			return 'notConnectedJetpack';
		}
		if ( this.checkProperty( 'isJetpackConnected' ) && this.checkProperty( 'userOwnsSite' ) ) {
			return 'alreadyConnected';
		}

		return false;
	}

	handleOnClickTos = () => this.props.recordTracksEvent( 'calypso_jpc_tos_link_click' );

	getTexts() {
		const { type, translate } = this.props;
		const selectedPlan = retrievePlan();

		if (
			type === 'pro' ||
			selectedPlan === 'jetpack_business' ||
			selectedPlan === 'jetpack_business_monthly'
		) {
			return {
				headerTitle: translate( 'Get Jetpack Professional' ),
				headerSubtitle: translate(
					'WordPress sites from start to finish: unlimited premium themes, ' +
						'business class security, and marketing automation.'
				),
			};
		}
		if (
			type === 'premium' ||
			selectedPlan === 'jetpack_premium' ||
			selectedPlan === 'jetpack_premium_monthly'
		) {
			return {
				headerTitle: translate( 'Get Jetpack Premium' ),
				headerSubtitle: translate(
					'Automated backups and malware scanning, expert priority support, ' +
						'marketing automation, and more.'
				),
			};
		}
		if (
			type === 'personal' ||
			selectedPlan === 'jetpack_personal' ||
			selectedPlan === 'jetpack_personal_monthly'
		) {
			return {
				headerTitle: translate( 'Get Jetpack Personal' ),
				headerSubtitle: translate(
					'Security essentials for every WordPress site ' +
						'including automated backups and priority support.'
				),
			};
		}

		if ( type === 'install' ) {
			return {
				headerTitle: translate( 'Install Jetpack' ),
				headerSubtitle: translate(
					'Jetpack brings free themes, security services, and essential marketing tools ' +
						'to your self-hosted WordPress site.'
				),
			};
		}
		return {
			headerTitle: translate( 'Connect a self-hosted WordPress' ),
			headerSubtitle: translate(
				"We'll be installing the Jetpack plugin so WordPress.com can connect to " +
					'your self-hosted WordPress site.'
			),
		};
	}

	isInstall() {
		/*
		 * FIXME: `return ! this.props.type` should be sufficient
		 * I'm avoiding significantly changing this implementation
		 * until propTypes have been around for a while.
		 */
		return includes( [ 'install', 'pro', 'premium', 'personal' ], this.props.type );
	}

	getInstructionsData( status ) {
		const { translate } = this.props;
		return {
			headerTitle:
				'notJetpack' === status
					? translate( 'Ready for installation' )
					: translate( 'Ready for activation' ),
			headerSubtitle: translate(
				"We'll need to send you to your site dashboard for a few manual steps."
			),
			steps:
				'notJetpack' === status
					? [ 'installJetpack', 'activateJetpackAfterInstall', 'connectJetpackAfterInstall' ]
					: [ 'activateJetpack', 'connectJetpack' ],
			buttonOnClick: 'notJetpack' === status ? this.installJetpack : this.activateJetpack,
			buttonText:
				'notJetpack' === status ? translate( 'Install Jetpack' ) : translate( 'Activate Jetpack' ),
		};
	}

	renderFooter() {
		const { translate } = this.props;
		return (
			<LoggedOutFormLinks>
				<JetpackConnectHappychatButton eventName="calypso_jpc_siteentry_chat_initiated">
					<LoggedOutFormLinkItem href="https://jetpack.com/support/installing-jetpack/">
						{ translate( 'Install Jetpack manually' ) }
					</LoggedOutFormLinkItem>
					{ this.isInstall() ? null : (
						<LoggedOutFormLinkItem href="/start">
							{ translate( 'Start a new site on WordPress.com' ) }
						</LoggedOutFormLinkItem>
					) }
					<HelpButton />
				</JetpackConnectHappychatButton>
			</LoggedOutFormLinks>
		);
	}

	renderSiteInput( status ) {
		return (
			<Card className="jetpack-connect__site-url-input-container">
				{ ! this.isCurrentUrlFetching() &&
				this.isCurrentUrlFetched() &&
				! this.props.jetpackConnectSite.isDismissed &&
				status ? (
					<JetpackConnectNotices
						noticeType={ status }
						onDismissClick={ this.dismissUrl }
						url={ this.state.currentUrl }
					/>
				) : null }

				<SiteUrlInput
					url={ this.state.shownUrl }
					onTosClick={ this.handleOnClickTos }
					onChange={ this.handleUrlChange }
					onSubmit={ this.handleUrlSubmit }
					isError={ this.getStatus() }
					isFetching={
						this.isCurrentUrlFetching() || this.isRedirecting() || this.state.waitingForSites
					}
					isInstall={ this.isInstall() }
				/>
			</Card>
		);
	}

	renderLocaleSuggestions() {
		if ( this.props.userModule.get() || ! this.props.locale ) {
			return;
		}

		return <LocaleSuggestions path={ this.props.path } locale={ this.props.locale } />;
	}

	renderSiteEntry() {
		const status = this.getStatus();
		return (
			<MainWrapper>
				{ this.renderLocaleSuggestions() }
				<div className="jetpack-connect__site-url-entry-container">
					<FormattedHeader
						headerText={ this.getTexts().headerTitle }
						subHeaderText={ this.getTexts().headerSubtitle }
					/>

					{ this.renderSiteInput( status ) }
					{ this.renderFooter() }
				</div>
			</MainWrapper>
		);
	}

	renderNotJetpackButton() {
		const { translate } = this.props;
		return (
			<a
				className="jetpack-connect__no-jetpack-button"
				href="#"
				onClick={ this.confirmJetpackNotInstalled }
			>
				{ translate( "Don't have jetpack installed?" ) }
			</a>
		);
	}

	renderBackButton() {
		const { translate } = this.props;
		return (
			<Button
				compact
				borderless
				className="jetpack-connect__back-button"
				onClick={ this.dismissUrl }
			>
				<Gridicon icon="arrow-left" size={ 18 } />
				{ translate( 'Back' ) }
			</Button>
		);
	}

	renderInstructions( instructionsData ) {
		const jetpackVersion = this.checkProperty( 'jetpackVersion' ),
			isInstall = this.isInstall(),
			{ currentUrl } = this.state;
		return (
			<MainWrapper isWide>
				{ this.renderLocaleSuggestions() }
				<div className="jetpack-connect__install">
					<FormattedHeader
						headerText={ instructionsData.headerTitle }
						subHeaderText={ instructionsData.headerSubtitle }
					/>
					<div className="jetpack-connect__install-steps">
						{ instructionsData.steps.map( ( stepName, key ) => {
							return (
								<JetpackInstallStep
									key={ 'instructions-step-' + key }
									stepName={ stepName }
									jetpackVersion={ jetpackVersion }
									isInstall={ isInstall }
									currentUrl={ currentUrl }
									confirmJetpackInstallStatus={ this.props.confirmJetpackInstallStatus }
									onClick={ instructionsData.buttonOnClick }
								/>
							);
						} ) }
					</div>
					<Button onClick={ instructionsData.buttonOnClick } primary>
						{ instructionsData.buttonText }
					</Button>
					<div className="jetpack-connect__navigation">{ this.renderBackButton() }</div>
				</div>
				<LoggedOutFormLinks>
					<JetpackConnectHappychatButton eventName="calypso_jpc_instructions_chat_initiated">
						<HelpButton />
					</JetpackConnectHappychatButton>
				</LoggedOutFormLinks>
			</MainWrapper>
		);
	}

	render() {
		const status = this.getStatus();
		if (
			includes( [ 'notJetpack', 'notActiveJetpack' ], status ) &&
			! this.props.jetpackConnectSite.isDismissed
		) {
			return this.renderInstructions( this.getInstructionsData( status ) );
		}
		return this.renderSiteEntry();
	}
}

const connectComponent = connect(
	state => ( {
		jetpackConnectSite: getConnectingSite( state ),
		getJetpackSiteByUrl: url => getJetpackSiteByUrl( state, url ),
		isRequestingSites: isRequestingSites( state ),
	} ),
	{
		confirmJetpackInstallStatus,
		recordTracksEvent,
		checkUrl,
		dismissUrl,
		goToRemoteAuth,
		goToPlans,
		goToPluginInstall,
		goToPluginActivation,
	}
);

export default flowRight( connectComponent, localize )( JetpackConnectMain );
