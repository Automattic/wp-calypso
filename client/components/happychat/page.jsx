import React from 'react';
import { connect } from 'react-redux';
import viewport from 'lib/viewport';
import { findDOMNode } from 'react-dom';

import { connectChat } from 'state/happychat/actions';
import { timeline, composer } from './helpers';

export const HappychatPage = React.createClass( {
	componentDidMount() {
		this.props.openChat();
	},

	onFocus() {
		const composerNode = findDOMNode( this.refs.composer );

		if ( viewport.isMobile() ) {
			/* User tapped textfield on a phone. This shows the keyboard. Unless we scroll to the bottom, the chatbox will be invisible */
			setTimeout( () => composerNode.scrollIntoView(), 500 );	/* Wait for the keyboard to appear */
		}
	},

	render() {
		const { connectionStatus } = this.props;
		return (
			<div className="happychat__container">
				{ timeline( { connectionStatus } ) }
				{ composer( { connectionStatus } ) }
			</div>
		);
	}
} );

const mapState = ( { happychat: { status: connectionStatus } } ) => ( { connectionStatus } );
const mapDispatch = dispatch => ( {
	openChat: () => dispatch( connectChat() )
} );
export default connect( mapState, mapDispatch )( HappychatPage );
