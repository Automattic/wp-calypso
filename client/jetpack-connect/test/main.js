/**
 * @format
 * @jest-environment jsdom
 */

/**
 * External dependencies
 */
import page from 'page';
import { externalRedirect } from 'lib/route';
import { noop } from 'lodash';

/**
 * Internal dependencies
 */
import { JetpackConnectMain } from '../main';

const REQUIRED_PROPS = {
	checkUrl: noop,
	confirmJetpackInstallStatus: noop,
	dismissUrl: noop,
	getJetpackSiteByUrl: noop,
	isRequestingSites: false,
	jetpackConnectSite: undefined,
	recordTracksEvent: jest.fn(),
};

jest.mock( 'page', () => ( {
	redirect: jest.fn(),
} ) );

jest.mock( 'lib/route/path', () => ( {
	externalRedirect: jest.fn(),
} ) );

describe( 'JetpackConnectMain', () => {
	describe( 'makeSafeRedirectionFunction', () => {
		const component = new JetpackConnectMain( REQUIRED_PROPS );

		beforeEach( () => {
			component.redirecting = false;
		} );

		test( 'should make a function that can calls the wrappe function', () => {
			const innerFunc = jest.fn();
			const wrapperFunc = component.makeSafeRedirectionFunction( innerFunc );
			expect( () => wrapperFunc() ).not.toThrow();
		} );

		test( 'should protect against multiple calls', () => {
			const innerFunc = jest.fn();
			const wrapperFunc = component.makeSafeRedirectionFunction( innerFunc );
			wrapperFunc();
			wrapperFunc();
			wrapperFunc();
			expect( innerFunc ).toHaveBeenCalledTimes( 1 );
		} );
	} );

	describe( 'goToPluginActivation', () => {
		const component = new JetpackConnectMain( REQUIRED_PROPS );

		beforeEach( () => {
			component.redirecting = false;
			component.props.recordTracksEvent.mockReset();
			externalRedirect.mockReset();
		} );

		test( 'should fire redirect', () => {
			component.goToPluginActivation( 'example.com' );

			expect( externalRedirect ).toHaveBeenCalledTimes( 1 );
			expect( externalRedirect.mock.calls[ 0 ] ).toMatchSnapshot();
		} );

		test( 'should dispatch analytics', () => {
			const url = 'example.com';
			component.goToPluginActivation( url );

			expect( component.props.recordTracksEvent ).toHaveBeenCalledTimes( 1 );
			expect( component.props.recordTracksEvent.mock.calls[ 0 ] ).toMatchSnapshot();
		} );
	} );

	describe( 'goToPluginInstall', () => {
		const component = new JetpackConnectMain( REQUIRED_PROPS );

		beforeEach( () => {
			component.redirecting = false;
			component.props.recordTracksEvent.mockReset();
			externalRedirect.mockReset();
		} );

		test( 'should fire redirect', () => {
			component.goToPluginInstall( 'example.com' );

			expect( externalRedirect ).toHaveBeenCalledTimes( 1 );
			expect( externalRedirect.mock.calls[ 0 ] ).toMatchSnapshot();
		} );

		test( 'should dispatch analytics', () => {
			const url = 'example.com';
			component.goToPluginInstall( url );

			expect( component.props.recordTracksEvent ).toHaveBeenCalledTimes( 1 );
			expect( component.props.recordTracksEvent.mock.calls[ 0 ] ).toMatchSnapshot();
		} );
	} );

	describe( 'goToRemoteAuth', () => {
		const component = new JetpackConnectMain( REQUIRED_PROPS );

		beforeEach( () => {
			component.redirecting = false;
			component.props.recordTracksEvent.mockReset();
			externalRedirect.mockReset();
		} );

		test( 'should fire redirect', () => {
			component.goToRemoteAuth( 'example.com' );

			expect( externalRedirect ).toHaveBeenCalledTimes( 1 );
			expect( externalRedirect.mock.calls[ 0 ] ).toMatchSnapshot();
		} );

		test( 'should dispatch analytics', () => {
			const url = 'example.com';
			component.goToRemoteAuth( url );

			expect( component.props.recordTracksEvent ).toHaveBeenCalledTimes( 1 );
			expect( component.props.recordTracksEvent.mock.calls[ 0 ] ).toMatchSnapshot();
		} );
	} );

	describe( 'goToPlans', () => {
		const component = new JetpackConnectMain( {
			...REQUIRED_PROPS,
			recordTracksEvent: jest.fn(),
		} );

		beforeEach( () => {
			component.redirecting = false;
			component.props.recordTracksEvent.mockReset();
			page.redirect.mockReset();
		} );

		test( 'should fire redirect', () => {
			component.goToPlans( 'example.com' );

			expect( page.redirect ).toHaveBeenCalledTimes( 1 );
			expect( page.redirect ).toHaveBeenCalledWith( '/jetpack/connect/plans/example.com' );
		} );

		test( 'should redirect to a site slug', () => {
			const url = 'https://example.com/sub-directory-install';
			component.goToPlans( url );

			expect( page.redirect ).toHaveBeenCalledWith(
				'/jetpack/connect/plans/example.com::sub-directory-install'
			);
		} );

		test( 'should dispatch analytics', () => {
			const url = 'example.com';
			component.goToPlans( url );

			expect( component.props.recordTracksEvent ).toHaveBeenCalledTimes( 1 );
			expect( component.props.recordTracksEvent.mock.calls[ 0 ] ).toMatchSnapshot();
		} );
	} );
} );
