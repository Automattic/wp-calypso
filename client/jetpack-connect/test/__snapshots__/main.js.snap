// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JetpackConnectMain goToPlans should dispatch analytics 1`] = `
Array [
  "calypso_jpc_success_redirect",
  Object {
    "type": "plans_selection",
    "url": "example.com",
  },
]
`;

exports[`JetpackConnectMain goToPluginActivation should dispatch analytics 1`] = `
Array [
  "calypso_jpc_success_redirect",
  Object {
    "type": "plugin_activation",
    "url": "example.com",
  },
]
`;

exports[`JetpackConnectMain goToPluginActivation should fire redirect 1`] = `
Array [
  "example.com/wp-admin/plugins.php?calypso_env=test",
]
`;

exports[`JetpackConnectMain goToPluginInstall should dispatch analytics 1`] = `
Array [
  "calypso_jpc_success_redirect",
  Object {
    "type": "plugin_install",
    "url": "example.com",
  },
]
`;

exports[`JetpackConnectMain goToPluginInstall should fire redirect 1`] = `
Array [
  "example.com/wp-admin/plugin-install.php?tab=plugin-information&plugin=jetpack&calypso_env=test",
]
`;

exports[`JetpackConnectMain goToRemoteAuth should dispatch analytics 1`] = `
Array [
  "calypso_jpc_success_redirect",
  Object {
    "type": "remote_auth",
    "url": "example.com",
  },
]
`;

exports[`JetpackConnectMain goToRemoteAuth should fire redirect 1`] = `
Array [
  "example.com/wp-admin/admin.php?page=jetpack&connect_url_redirect=true&calypso_env=test",
]
`;
